quinn-proto/src/connection/assembler.rs

```diff
@@ -152,46 +152,58 @@ impl Assembler {
                 }
             } else {
                 let bytes = buffer.split().freeze();
-                self.data.push(Chunk { offset, bytes });
+                self.data.push(Chunk {
+                    offset,
+                    bytes,
+                    over_allocation: 0,
+                });
                 offset = chunk.offset;
                 buffer.extend_from_slice(&chunk.bytes);
             }
         }
 
         let bytes = buffer.split().freeze();
-        self.data.push(Chunk { offset, bytes });
-        self.defragmented = self.data.len();
+        self.data.push(Chunk {
+            offset,
+            bytes,
+            over_allocation: 0,
+        });
+        self.over_allocation = 0;
     }
 
-    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes) {
+    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes, mut allocation_size: usize) {
         self.end = self.end.max(offset + bytes.len() as u64);
-
         if let State::Unordered { ref mut recvd } = self.state {
+            let start = offset;
             // Discard duplicate data
             for duplicate in recvd.replace(offset..offset + bytes.len() as u64) {
                 if duplicate.start > offset {
+                    let over_allocation = (duplicate.end - duplicate.start) as usize;
                     self.data.push(Chunk {
                         offset,
                         bytes: bytes.split_to((duplicate.start - offset) as usize),
+                        over_allocation,
                     });
+                    self.over_allocation += over_allocation;
                     offset = duplicate.start;
                 }
                 bytes.advance((duplicate.end - offset) as usize);
                 offset = duplicate.end;
             }
+            allocation_size = allocation_size.saturating_sub((offset - start) as usize);
```

Comment:
I'm not sure about this. As far as I understand it, either `Chunk` could keep the entire allocation alive. It's also already the case that a payload may have multiple `STREAM` frames (either for the same stream or for different streams), so we're generally somewhat overcounting the overallocations, and we just have to take that into account when deciding to defragment (but to that end we can just fudge the defragmentation trigger heuristic).

Reply:
I still would like the over-allocation estimate to be correct and precise if the `allocation_size` argument is also correct and precise. 
So e.g. `insert(1, Buffer of size 1022, 1022)` followed by `insert(0, Buffer of size 1024, 1024)` should increase over-allocation by 1022 and not by 2045.
I agree that we will often over-count the over-allocation, but I don't see that as a reason to make the estimate even worse.

Reply:
That's not really my point though. My point is this: shrinking the `Bytes` doesn't shrink the size of the allocation, it simply changes the index pointing to the allocation. So in that sense, the deduplication happening here should have, in my understanding, no effect on the amount of overallocation.

Reply:
You are right that only advancing the `Bytes` should not shrink the `allocation_size`, but pushing a `Buffer` should, because we do not want to account for the same bytes twice. My error was that I thought that I could simply adjust the `allocation_size` at the end of the de-duplication loop, but now I see that this does not work when we are not pushing a new `Buffer` at every iteration step. I therefore moved the `allocation_size` adjustment into the loop, only adjusting it when we are pushing a new `Buffer`.

Reply:
But shrinking the allocation size is only warranted if we're accounting for the same *allocations*. A different but overlapping `Buffer` ("same bytes twice") is substantially more likely to originate from a different packet, and thus from a different allocation.

Reply:
I try to clarity my thoughts a bit more:

```rust
if duplicate.start > offset {
    let over_allocation = (duplicate.end - duplicate.start) as usize;
    self.data.push(Buffer {
        offset,
        bytes: bytes.split_to((duplicate.start - offset) as usize),
        over_allocation,
    });
    self.over_allocation += over_allocation;
    allocation_size =
        allocation_size.saturating_sub((duplicate.end - offset) as usize);
    offset = duplicate.start;
}
```
Here we are pushing a new `Buffer`, because we need to close the gap to the next duplicate. This new `Buffer` references the same allocation and I use the size of the duplicate as the over-allocation for the new `Buffer`, because this is the amount of bytes that are unused but are still referenced and kept in memory by the `Buffer`. The reasons I associate a over-allocation with every `Buffer` are:
- It could be the last `Buffer` we push
- If we only keep track of over-allocation in a single `Buffer`, when that `Buffer` is later removed, we will underestimate the over-allocation caused by the remaining Buffers pushed by the `insert` call.

Because I account for the over-allocation in the pushed `Buffer`, I reduce the allocation size by the size of the `Buffer` and its over-allocation.

A possible other sensible approach would be to calculate for every pushed Buffer the over-allocation as `allocation_size.saturating_sub(bytes.len())` and never mutating the `allocation_size`. Then the  calculated over-allocation would grow much larger if we have to fill many small gaps in the de-duplication loop.